server/src/main/scala/com/cloudera/livy/server/batch/BatchSession.scala

```diff
@@ -23,62 +23,129 @@ import java.lang.ProcessBuilder.Redirect
 import scala.concurrent.{ExecutionContext, ExecutionContextExecutor}
 import scala.util.Random
 
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties
+
 import com.cloudera.livy.LivyConf
+import com.cloudera.livy.server.recovery.SessionStore
 import com.cloudera.livy.sessions.{Session, SessionState}
+import com.cloudera.livy.sessions.Session.RecoveryMetadata
+import com.cloudera.livy.util.LineBufferedProcess
 import com.cloudera.livy.utils.{AppInfo, SparkApp, SparkAppListener, SparkProcessBuilder}
 
+@JsonIgnoreProperties(ignoreUnknown = true)
+case class BatchRecoveryMetadata(
+    id: Int,
+    appTag: String,
+    appId: Option[String],
+    owner: String,
+    proxyUser: Option[String],
+    version: Int = 1)
+  extends RecoveryMetadata
+
+object BatchSession {
+  val RECOVERY_SESSION_TYPE = "batch"
+
+  def create(
+      id: Int,
+      owner: String,
+      proxyUser: Option[String],
+      livyConf: LivyConf,
+      createRequest: CreateBatchRequest,
+      sessionStore: SessionStore,
+      mockApp: Option[SparkApp] = None): BatchSession = {
+    val appTag = s"livy-batch-$id-${Random.alphanumeric.take(8).mkString}"
+    new BatchSession(
+      id, appTag, owner, proxyUser, livyConf, Left(createRequest), sessionStore, mockApp)
+  }
+
+  def recover(
+      m: BatchRecoveryMetadata,
+      livyConf: LivyConf,
+      sessionStore: SessionStore,
+      mockApp: Option[SparkApp] = None): BatchSession = {
+    new BatchSession(
+      m.id, m.appTag, m.owner, m.proxyUser, livyConf, Right(m.appId), sessionStore, mockApp)
+  }
+}
+
 class BatchSession(
```

Comment:
I totally agree and will fix.
However, I'm a bit confused. In the batch-yarn you mentioned that you don't consider singleton object as a part of the class and asked me to move those logic from the singleton factory to the class. And now we are moving the logic back to the singleton factory.